subsys/fs/littlefs_fs.c

```diff
@@ -0,0 +1,730 @@
+		return -EBUSY;
+	}
+
+	/* Create and take mutex. */
+	k_mutex_init(&fs->mutex);
+	fs_lock(fs);
+
+	/* Open flash area */
+	ret = flash_area_open(area_id, &fs->area);
+	if ((ret < 0) || (fs->area == NULL)) {
+		LOG_ERR("can't open flash area %d", area_id);
+		ret = -ENODEV;
+		goto out;
+	}
+	LOG_DBG("FS area %u at 0x%x for %u bytes",
+		area_id, (u32_t)fs->area->fa_off,
+		(u32_t)fs->area->fa_size);
+
+	dev = flash_area_get_device(fs->area);
+	if (dev == NULL) {
+		LOG_ERR("can't get flash device: %s", fs->area->fa_dev_name);
+		ret = -ENODEV;
+		goto out;
+	}
+
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_READ_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_PROG_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_CACHE_SIZE > 0);
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE > 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE % 8) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
+		      % CONFIG_FS_LITTLEFS_READ_SIZE) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
+		      % CONFIG_FS_LITTLEFS_PROG_SIZE) == 0);
+
+	struct lfs_config *lcp = &fs->cfg;
+
+	lfs_size_t read_size = lcp->read_size;
+
+	if (read_size == 0) {
+		read_size = CONFIG_FS_LITTLEFS_READ_SIZE;
+	}
+
+	lfs_size_t prog_size = lcp->prog_size;
+
+	if (prog_size == 0) {
+		prog_size = CONFIG_FS_LITTLEFS_PROG_SIZE;
+	}
+
+	/* Yes, you can override block size. */
+	lfs_size_t block_size = lcp->block_size;
+
+	if (block_size == 0) {
+		block_size = get_block_size(fs->area);
+	}
+
+	s32_t block_cycles = lcp->block_cycles;
+
+	if (block_cycles == 0) {
+		block_cycles = CONFIG_FS_LITTLEFS_BLOCK_CYCLES;
+	}
+	if (block_cycles <= 0) {
+		/* Disable leveling (littlefs v2.1+ semantics) */
+		block_cycles = -1;
+	}
+
+	lfs_size_t cache_size = lcp->cache_size;
+
+	if (cache_size == 0) {
+		cache_size = CONFIG_FS_LITTLEFS_CACHE_SIZE;
+	}
+
+	lfs_size_t lookahead_size = lcp->lookahead_size;
+
+	if (lookahead_size == 0) {
+		lookahead_size = CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE;
+	}
+
+
+	/* No, you don't get to override this. */
+	lfs_size_t block_count = fs->area->fa_size / block_size;
+
+	LOG_INF("FS at %s:0x%x is %u 0x%x-byte blocks with %u cycle",
+		dev->config->name, (u32_t)fs->area->fa_off,
+		block_count, block_size, block_cycles);
+	LOG_INF("sizes: rd %u ; pr %u ; ca %u ; la %u",
+		read_size, prog_size, cache_size, lookahead_size);
+
+	__ASSERT_NO_MSG(prog_size != 0);
+	__ASSERT_NO_MSG(read_size != 0);
+	__ASSERT_NO_MSG(cache_size != 0);
+	__ASSERT_NO_MSG(block_size != 0);
+
+	__ASSERT((fs->area->fa_size % block_size) == 0,
+		 "partition size must be multiple of block size");
+	__ASSERT((block_size % prog_size) == 0,
+		 "erase size must be multiple of write size");
+	__ASSERT((block_size % cache_size) == 0,
+		 "cache size incompatible with block size");
+
+	/* Set the validated/defaulted values. */
+	lcp->context = (void *)fs->area;
+	lcp->read = lfs_api_read;
+	lcp->prog = lfs_api_prog;
+	lcp->erase = lfs_api_erase;
+	lcp->sync = lfs_api_sync;
+	lcp->read_size = read_size;
```

Comment:
@pkral78 Here is where `lfs_config` must be mutable.  The pointers to the device-private functions get assigned, as well as information about where in the flash device the specific partition is located.

Reply:
No it does not need to, those are just immutable function pointers or constants known at compile time. I see block size is get from `get_block_size` call, however I don't understand why it need to be calculated at runtime, when it will be always same for same target configuration.

Reply:
`prog`, `erase`, and `sync` are known at compile time, but they're private to this module.  How do you expect to be able to set them at compile time in const struct stored in flash and declared in a different module?

The flash characteristics are determined at runtime because Zephyr's devicetree implementation only gives you the flash area identifier.  The remaining information simply isn't available readily at compile time.

Reply:
> `prog`, `erase`, and `sync` are known at compile time, but they're private to this module. How do you expect to be able to set them at compile time in const struct stored in flash and declared in a different module?
> 

And does need to be private ? 

> The flash characteristics are determined at runtime because Zephyr's devicetree implementation only gives you the flash area identifier. The remaining information simply isn't available readily at compile time.

Yes you're right, but it's rather DT limitation (an area for improvement?).

I think that these two questions lead to decision if we could accept this PR as is and left my comments for future improvements. I am for accepting this PR as it is super useful and well written.